### PR TITLE
feat(fastify): support fastify as plugin

### PR DESCRIPTION
**Pull request by tdeekens**
#### Summary

Release coordination PR for fasitfy integration.

### CODE REVIEW

File: packages/fastify/modules/index.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+const { version } = require('../package.json');
+const {
+  plugin,
```
Comment (tdeekens) [inline packages/fastify/modules/index.js]: Renamed this to be inline with the other plugins/middlewares. The idea is that when somebody imports it should be `const { plugin: promsterPlugin } = require('@promster/fastify')`.

File: packages/fastify/modules/plugin/plugin.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
+const fastifyPlugin = require('fastify-plugin');
```
Comment (tdeekens) [inline packages/fastify/modules/plugin/plugin.js]: Renamed.

File: packages/fastify/modules/plugin/plugin.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        // eslint-disable-next-line camelcase
+        status_code: defaultedOptions.normalizeStatusCode(reply.res.statusCode),
+        path: defaultedOptions.normalizePath(extractPath(req)),
+        ...(defaultedOptions.getLabelValues &&
+          defaultedOptions.getLabelValues(req, reply)),
+      },
+    });
+  });
+
+  fastify.route({
```
Comment (tdeekens) [inline packages/fastify/modules/plugin/plugin.js]: This I will remove. I am aware that this is a bit controversal but I would like to 

1. Enforce best practices: the Prometheus server should be another server aside the app
   - This makes it run on another port (when using `@promster/server`) and makes it easier to not expose in a k8s setup
2. Give people more flexibility has to how to expose it with very few lines of code if they want to

File: packages/fastify/modules/plugin/plugin.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+let recordRequest;
+let upMetric;
+
+const extractPath = req => req.raw.originalUrl || req.raw.url;
+const getRequestRecorder = () => recordRequest;
+const signalIsUp = () => upMetric && upMetric.set(1);
+const signalIsNotUp = () => upMetric && upMetric.set(0);
+
+const createPlugin = async (fastify, options) => {
```
Comment (tdeekens) [inline packages/fastify/modules/plugin/plugin.js]: Renamed.

**Comment by dimonnwc3** [comment]
@tdeekens looks good.
One more idea we can attach up/down status to fastify hooks as well
like calling `up(1)` on plugin init and `up(0)` on `onClose` [hook](https://github.com/fastify/fastify/blob/master/docs/Hooks.md#application-hooks)

but here are several thoughts:
`up(1)` on plugin init - doesn't guarantee that app will boot up successfully
especially when we expose Prometheus metrics on a separate port it may cause inconsistent state where:
- server for Prometheus is running and returns `up(1)` status
- main app server is failed by some reason and not responding at all

**Comment by tdeekens** [comment]
> One more idea we can attach up/down status to fastify hooks as well

Yes, thanks for the idea and thoughts. They reflect the same reasoning I had why I didn't add those into hooks or plugin init in other packages. It gives more control when bootstrapping your server which may require other things: loading config or checking or other upstream services. All of which may give false metrics which can cascade through infra in annying/hard to spot ways. That's why I also wouldn't do that for now.

**Comment by tdeekens** [comment]
Published: @promster/fastify: 1.0.0 => 1.0.0.

**Pull request merged**